workflow/scripts/osemosys_global/powerplant/data.py

```diff
@@ -0,0 +1,334 @@
+    df_dict_2 = pd.merge(df_dict_fuel, df_dict_nodes, how="outer", on="powerplant")
+    
+    
+    ## Merge original generator dataframe with nodes and fuels
+    df_gen_base = pd.merge(df_gen_base, df_dict_2, how="outer", on="powerplant")
+    df_gen_base.rename(
+        {"child_object_x": "fuel", "child_object_y": "node"}, axis=1, inplace=True
+    )
+    
+    ## Extract start year from Commission Date
+    df_gen_base["Commission Date"] = (pd.to_timedelta(df_gen_base["Commission Date"].astype(int),
+                                                unit='d') + 
+                               datetime(1900, 1, 1))
+    
+    df_gen_base["start_year"] = df_gen_base["Commission Date"].dt.year
+    df_gen_base.drop("Commission Date", axis=1, inplace=True)
+    
+    ## Calculate efficiency from heat rate. Units of heat rate in MJ/kWh
+    df_gen_base["efficiency"] = 3.6 / df_gen_base["Heat Rate"].astype(float)
+    df_gen_base.drop("Heat Rate", axis=1, inplace=True)
+    
+    ## Calcluate years of operation from start year
+    df_gen_base["years_of_operation"] = start_year - df_gen_base["start_year"]
+    
+    ## Fix blank spaces in 'fuels' columns. Appearing for 'Oil' powerplants in certain countries
+    df_gen_base.loc[df_gen_base["fuel"].isna(), "fuel"] = (
+        df_gen_base["node"].str.split("-").str[:2].str.join("-")
+        + " "
+        + df_gen_base["powerplant"].str.split("_", expand=True)[1]
+    )
+
+    ## Create column for technology
+    df_gen_base["technology"] = df_gen_base["powerplant"].str.split("_").str[1]
+    df_gen_base["technology"] = df_gen_base["technology"].str.title()
+
+    ## Divide Gas into CCGT and OCGT based on max capacity
+    df_gen_base.loc[
+        (df_gen_base["technology"] == "Gas") & (df_gen_base["Max Capacity"].astype(float) > 130),
+        "technology",
+    ] = "Gas-CCGT"
+    df_gen_base.loc[
+        (df_gen_base["technology"] == "Gas") & (df_gen_base["Max Capacity"].astype(float) <= 130),
+        "technology",
+    ] = "Gas-OCGT"
+
+    # Create table with aggregated capacity  
+    df_gen_agg_node = df_gen_base[df_gen_base['start_year']<=start_year]
+    df_gen_agg_node = df_gen_agg_node.groupby(['node', 'technology'], 
+                                              as_index=False)['total_capacity'].sum()
+    df_gen_agg_node = df_gen_agg_node.pivot(index='node', 
+                                            columns='technology', 
+                                            values='total_capacity').fillna(0).reset_index()
+
+    df_gen_agg_node.drop('Sto', axis=1, inplace=True) # Drop 'Sto' technology. Only for USA.
+        
+    nodes_extra_df = pd.DataFrame(columns=['node'])
+    
+    nodes_extra_df['node'] = nodes_extra_list
+
+    df_gen_agg_node = pd.concat(
+        [df_gen_agg_node,nodes_extra_df],
+        ignore_index=True,
+        sort=False,
+    ).fillna(0).sort_values(by='node').set_index('node').round(2)
```

Comment:
Is this doing anything? I seem to get the same results if I dont include the extra nodes?

Reply:
Not sure, it's legacy code. Did you try it with a geographical scope where one of the added nodes is included from the NODES_EXTRA_LIST constant?